library/ecp_curves.c

```diff
@@ -5527,9 +5534,15 @@ int mbedtls_ecp_mod_p448(mbedtls_mpi_uint *X, size_t X_limbs)
         X[i] = 0;
     }
 
-    /* X += A1 - Carry here dealt with by oversize M and X. */
+    /* X += A1 - Carry here fits in oversize X. Oversize M means it will get
+     * added in, not returned as carry. */
     (void) mbedtls_mpi_core_add(X, X, M, M_limbs);
 
+    /* Deal with carry bit from add by subtracting P if necessary. */
+    if (X[P448_WIDTH] != 0) {
+        mbedtls_mpi_core_sub(X, X, P, P_limbs);
+    }
+
```

Comment:
Does that mean: "simply subtract the modulu P" can't ensure the modulo function work correctly? or will it need more than one time subtraction?

Reply:
I also added some comments that I found useful, please feel free to take this commit, if you too find it helpful: https://github.com/yanesca/mbedtls/commit/9873dad8e8c0e3d13c68e58b090fd97d3b967c04

Reply:
After a second thought: there is one more fixup missing. We basically need to process the carry after every addition. And that means that we are not making much of a difference in terms of performance.

@xkqian I think it would be better to implement @gabor-mezei-arm 's suggestion. That is, we should remove the parts where we conditionally deduct P and accumulate the carry in X and do another round of reduction in the end. (In a straightforward implementation it would mean adding a loop that does the reduction exactly two times.)

Suggested optimisations:
- replace zeroising loops with calls to memset
- replace the `shift_r` with memcpy as Gabor [suggested](https://github.com/Mbed-TLS/mbedtls/pull/7566#discussion_r1200620577)
- instead of `shift_l` just adjust the pointer that we pass to `mbedtls_mpi_core_add()` afterwards (safe to skip the lower limbs as we are adding 0 there and only use addition on the higher limbs)
- in the second round A1 and B0 have at most 1 non-zero chunk and B1=0. Using this we need to calculate A0 + A1 + B1 + (B0 + B1) * 2^224 = A0 + A1 + B0 * 2^224. Here we can construct A1 + B0 * 2^224 in a temporary directly without any call to `mbedtls_mpi_core_add()` (both of them are all zero except in a single, non-colliding limb). This means, the single call to `mbedtls_mpi_core_add()` we need to make will calculate the end result A0 + (A1 + B0 * 2^224)

Reply:
Some more thoughts:
- After the main round, we need two additional reduction rounds to make sure that we don't have any carry (one additional reduction round is not enough)
- For maintainability and code size we can just put the final round into a loop that executes exactly twice

Reply:
> * After the main round, we need two additional reduction rounds to make sure that we don't have any carry (one additional reduction round is not enough)
> * For maintainability and code size we can just put the final round into a loop that executes exactly twice

I am sorry, I am trying to understand the description, just to confirm: Does "`does the reduction exactly two three times"` include "**main round**" and "`two additional reduction round`" here?

My understanding here, after first reduction round(A0 + A1 + B1 + (B0 + B1) * 2^224) ①, the updated X will have at most two bits none-zero A1, that is: B1: total zero, and B0: at most 2 bits.   So the second round reduction will have the formula like: (A0 + A1 + 0 + (B0 + 0) * 2^224) ②-> A0 + (B0 + B0*2^224) ③. cause B0 has at most 2 bits. and the sum of  ③ has at most 1 bit exceeded, So whether the last round can be implemented by simply deduct the modulus?

Reply:
> Does "does the reduction exactly two three times" include "main round" and "two additional reduction round" here?

Yes, it includes the "main" round and two additional rounds.

> My understanding here, after first reduction round(A0 + A1 + B1 + (B0 + B1) * 2^224) ①, the updated X will have at most two bits none-zero A1, that is: B1: total zero, and B0: at most 2 bits. So the second round reduction will have the formula like: (A0 + A1 + 0 + (B0 + 0) * 2^224) ②-> A0 + (B0 + B0*2^224) ③.

Did you mean A0 + A1 + B0*2^224 for ③ ? Other than that, yes that is correct.

> and the sum of ③ has at most 1 bit exceeded, So whether the last round can be implemented by simply deduct the modulus?

Indeed, ③ has at most 1 bit exceeded and we could potentially just deduct the modulus. The reason why I was suggesting to do another round is twofold:
- If we do another round it can be done with a loop, which is better for code size and maintenance.
- When we make the code constant time, deducting the modulus will be equivalent to two bignum additions, whereas repeating the second round takes only one. This means that repeating the second round is not worse and will eventually be better in terms of performance as well.

Reply:
> > My understanding here, after first reduction round(A0 + A1 + B1 + (B0 + B1) * 2^224) ①, the updated X will have at most two bits none-zero A1, that is: B1: total zero, and B0: at most 2 bits. So the second round reduction will have the formula like: (A0 + A1 + 0 + (B0 + 0) * 2^224) ②-> A0 + (B0 + B0*2^224) ③.
> 
> Did you mean A0 + A1 + B0*2^224 for ③ ? Other than that, yes that is correct.
 Yes,  I mean  **A0 + A1 + B0*2^224** , cause  A1 = B1*2^224 + B0, and B1=0,  we can know: A1= B0, take it ino ②, then we can get A0 + (B0 + B0*2^224) ③
> 
> > and the sum of ③ has at most 1 bit exceeded, So whether the last round can be implemented by simply deduct the modulus?
> 
> Indeed, ③ has at most 1 bit exceeded and we could potentially just deduct the modulus. The reason why I was suggesting to do another round is twofold:
> 
> * If we do another round it can be done with a loop, which is better for code size and maintenance.
> * When we make the code constant time, deducting the modulus will be equivalent to two bignum additions, whereas repeating the second round takes only one. This means that repeating the second round is not worse and will eventually be better in terms of performance as well.

Got it. Thanks for the explanation.